### PR TITLE
add .pub-cache back to .gitignore

### LINKED ISSUE #128635: Flutter upgrade ko for .pub-cache folder
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Install an older flutter version (E.g. 3.10.3)
2. run `flutter pub get` command inside a flutter project with dependencies
3. run flutter upgrade command

### Expected results

The upgrade starts without error and upgrade flutter to latest stable version ignoring (or deleting?) `.pub-cache/` folder.

### Actual results

The upgrade process was stopped by git check, because a change inside local repository was found.

```
$ flutter upgrade
Your flutter checkout has local changes that would be erased by upgrading. If you want to keep these changes, it is recommended that you stash them via "git stash" or else commit the changes to a local branch.
If it is okay to remove local changes, then re-run this command with "--force".
```

The problem is related to `.pub-cache/` folder inside flutter directory:

```
~/snap/flutter/common/flutter$ git status
On branch stable
Your branch is behind 'origin/stable' by 1 commit, and can be fast-forwarded.
  (use "git pull" to update your local branch)

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	.pub-cache/

nothing added to commit but untracked files present (use "git add" to track)
```

### Code sample

No code related issue

### Screenshots or Video

_No response_

### Logs

_No response_

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
$ flutter doctor -v

┌─────────────────────────────────────────────────────────┐
│ A new version of Flutter is available!                  │
│                                                         │
│ To update to the latest version, run "flutter upgrade". │
└─────────────────────────────────────────────────────────┘
[✓] Flutter (Channel stable, 3.10.3, on Ubuntu 22.04.2 LTS 5.19.0-43-generic, locale en_US.UTF-8)
    • Flutter version 3.10.3 on channel stable at /home/vincenzosalerno/snap/flutter/common/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f92f44110e (8 days ago), 2023-06-01 18:17:33 -0500
    • Engine revision 2a3401c9bb
    • Dart version 3.0.3
    • DevTools version 2.23.1

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.


[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).

[✓] IntelliJ IDEA Community Edition (version 2023.1)
    • IntelliJ at /snap/intellij-idea-community/434
    • Flutter plugin version 74.0.4
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.78.2)
    • VS Code at /usr/share/code
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 22.04.2 LTS 5.19.0-43-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 114.0.5735.106

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.
```

</details>

### PR DESCRIPTION
**Pull request by christopherfujino**
Fixes https://github.com/flutter/flutter/issues/128635

In https://github.com/flutter/flutter/commit/1fd84f88e915427c88fb1d38c052150c76e27b4f, we renamed the `.pub-cache` entry in the `.gitignore` file to `.pub-preload-cache`, and ensured the packaging script was now populating that directory. However, for users who already had this directory from downloading an earlier archive but used `flutter upgrade` to get the latest code, they now had a `.gitignore` file that was no longer ignoring their `.pub-cache` directory.

Now, when these users try to `flutter upgrade` again, the tool tries to verify if their checkout is clean, and will exit early if not directing the user to either stash or commit these changes, or do `flutter upgrade --force` which will try to update their branch anyway (and which would succeed since there would be no git conflict). These change adds back the `.pub-cache` entry to `.gitignore`, which won't retroactively fix broken releases, but will at least ensure if they `flutter upgrade --force` once to get this fix, they won't need to again.

### CODE REVIEW

**Comment by Hixie** [comment]
test-exempt: configuration change

but bonus points for actually testing it. obviously the simple test is just literally checking the file contains what you expect, but the more elaborate test would be checking that the scenario you describe works as expected.

**Comment by christopherfujino** [comment]
> test-exempt: configuration change
> 
> but bonus points for actually testing it. obviously the simple test is just literally checking the file contains what you expect, but the more elaborate test would be checking that the scenario you describe works as expected.

I considered writing a test, but I'm not sure there would be much value. I'm not too worried about `.pub-cache` in particular regressing, ideally I would have a test that verifies we never remove anything from `.gitignore`, only add things (with a special allow-list for special cases where handle deleting the file from checkouts before/after upgrading).

However we don't currently have infra for seeing the diff associated with the current PR (we can't check the current commit because there may be multiple commits in a PR, and the HEAD commit may not have the offending change).

**Comment by christopherfujino** [comment]
> the more elaborate test would be checking that the scenario you describe works as expected.

This would be difficult to do. Ideally you would download a prebuilt archive and then `flutter upgrade` to the current commit you're testing, but then that feature branch commit won't exist upstream. And I think doing something like `git remote add ...` is a bad idea, and would break, as it's quite complicated all the places we fetch commits from across our CI processes.

Also we'd have the problem of which older archive to grab. The perfect test would grab all published release archives ever and ensure they can successfully upgrade to the commit under test.

**Comment by andrewkolos** [comment]
Would it be worth adding a comment to the .gitignore cautioning against about removing entries?

**Review by andrewkolos — COMMENTED** [review]
See previous comment (I forgot to mark it as a review)

**Comment by christopherfujino** [comment]
> See previous comment (I forgot to mark it as a review)

Done

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,6 @@
+# Do not remove or rename entries in this file, only add new ones
+# See https://github.com/flutter/flutter/issues/128635 for more context
```
Comment (andrewkolos) [inline .gitignore:2]:
```suggestion
# See https://github.com/flutter/flutter/issues/128635 for more context.
```

**Review by andrewkolos — APPROVED** [review]
LGTM with uber nit

**Comment by sigurdm** [comment]
> In https://github.com/flutter/flutter/commit/1fd84f88e915427c88fb1d38c052150c76e27b4f, we renamed the .pub-cache entry in the .gitignore file to .pub-preload-cache

Just a small correction for later context. This was not a rename:
* The FLUTTER_ROOT/.pub_cache was an actual pub cache that `flutter pub` would use 
* The `.pub-preload-cache` contains package archives, and on first run they are extracted to the user-level pub-cache (eg ~/.pub_cache on mac).

**Comment by christopherfujino** [comment]
> > In [1fd84f8](https://github.com/flutter/flutter/commit/1fd84f88e915427c88fb1d38c052150c76e27b4f), we renamed the .pub-cache entry in the .gitignore file to .pub-preload-cache
> 
> Just a small correction for later context. This was not a rename:
> 
> * The FLUTTER_ROOT/.pub_cache was an actual pub cache that `flutter pub` would use
> * The `.pub-preload-cache` contains package archives, and on first run they are extracted to the user-level pub-cache (eg ~/.pub_cache on mac).

Understood. I meant the effective change to the .gitignore, but I should have been more clear.

**Pull request merged**